tree.go

```diff
@@ -1293,3 +1266,155 @@ func ToDot(root VerkleNode) string {
 	root.Commit()
 	return fmt.Sprintf("digraph D {\n%s}", root.toDot("", ""))
 }
+
+type SerializedNode struct {
+	Node            VerkleNode
+	CommitmentBytes [32]byte
+	SerializedBytes []byte
+}
+
+// BatchSerialize is an optimized serialization API when multiple VerkleNodes serializations are required, and all are
+// available in memory.
+func (n *InternalNode) BatchSerialize() ([]SerializedNode, error) {
+	// Commit to the node to update all the nodes commitments.
+	n.Commit()
+
+	// Collect all nodes that we need to serialize.
+	nodes := make([]VerkleNode, 0, 1024)
+	nodes = n.collectNonHashedNodes(nodes)
+
+	// We collect all the *Point so we can batch invert the work the projective->affine transformation.
+	pointsToCompress := make([]*Point, 0, len(nodes)*NodeWidth+len(nodes))
+	pointToCompressPerNodeCount := make([]int, len(nodes))
+	for i := range nodes {
+		switch n := nodes[i].(type) {
+		case *InternalNode:
+			pointsToCompress = append(pointsToCompress, n.commitment)
+			count := 0
+			for _, c := range n.children {
+				_, isInternalNode := c.(*InternalNode)
+				_, isLeafNode := c.(*LeafNode)
+				if isInternalNode || isLeafNode {
+					pointsToCompress = append(pointsToCompress, c.Commitment())
+					count++
+				}
+			}
+			pointToCompressPerNodeCount[i] = 1 + count // Internal node commitment + all non-hashed children.
+		case *LeafNode:
+			pointsToCompress = append(pointsToCompress, n.commitment, n.c1, n.c2)
+			pointToCompressPerNodeCount[i] = 3 // Leaf node commitment + c1 + c2.
+		}
+	}
+
+	// Now we do the all transformations in a single-shot.
+	compressedPoints := banderwagon.ElementsToBytes(pointsToCompress)
+
+	// Now we that we did the heavy CPU work, we have to do the rest of `nodes` serialization
+	// taking the compressed points from this single list.
+	ret := make([]SerializedNode, 0, len(nodes))
+	for i := range nodes {
+		switch n := nodes[i].(type) {
+		case *InternalNode:
+			compressedChildren := compressedPoints[1:pointToCompressPerNodeCount[i]]
+			sn := SerializedNode{
+				Node:            n,
+				CommitmentBytes: compressedPoints[0],
+				SerializedBytes: n.serializeWithCompressedChildren(compressedChildren),
+			}
+			ret = append(ret, sn)
+		case *LeafNode:
+			c1Bytes := compressedPoints[0]
+			c2Bytes := compressedPoints[1]
+			sn := SerializedNode{
+				Node:            n,
+				CommitmentBytes: compressedPoints[0],
+				SerializedBytes: n.serializeWithCompressedCommitments(c1Bytes, c2Bytes),
+			}
+			ret = append(ret, sn)
+		}
+		compressedPoints = compressedPoints[pointToCompressPerNodeCount[i]:]
+	}
+
+	return ret, nil
+}
+
+func (n *InternalNode) collectNonHashedNodes(list []VerkleNode) []VerkleNode {
+	list = append(list, n)
+	for _, child := range n.children {
```

Comment:
see my 2nd question in the general review summary: would it be possible to use `n.cow` here?

Reply:
Uhm, unfortunately it looks that it won't work in that case and I think I know the reason.

The Verkle Trie is called `Commit()` before the `BatchSerialize()` call in go-ethereum. It first does [trie.Hash()](https://github.com/gballet/go-ethereum/blob/verkle/replay-change-with-tree-group-tryupdate/core/state/statedb.go#L979), which does `tire.root.Commit().Bytes()`... so unfortunately at the start of `BatchSerialize()` we already lost the `cows`.

This for me opens an even bigger question that probably you'll know better. That `IntermediateRoot` is committing _per transaction and not block-wise? If that's the case, the new batching that we do for `Commit()` is batching trie changes per txn, and not per block (thus smaller). Is that fine? (just double checking)

What this means for serialization here, is that when `BatchSerialize(..)` is called, we already lost all the `cows` already so we can't use them for anything as suggested.

[Maybe we can think of a way to keep some sort of "cow" that lives from the trie opening until the serialization or something like that. But I'd like first to get your thoughts before going deeper on potential solutions]

Reply:
Circling this a bit more looks like our replay benchmark doesn't detect our "past byzantium" hard fork, which I remember you mentioned some months ago this "intermediate root" calculation was removed.

I [double checked here](https://github.com/gballet/go-ethereum/blob/7d1fb48778c3e2b93913738708c66d5e5bf8f2ec/core/state_processor.go#L114), and in the replay run it takes the `else` route.

So this is a bit pessimistic for our intentions since the more we can batch, the better our optimizations would work compared with MPT (which probably doesn't care that much about "batching" as VKT crypto).